torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -309,36 +311,52 @@ void TensorPipeAgent::startImpl() {
         priority, std::move(key), std::move(reg->transport));
   }
 
-
-  for (auto& key : TensorPipeCpuChannelRegistry()->Keys()) {
-    int64_t priority = -1;
-    if (opts_.channels.has_value()) {
-      auto iter =
-          std::find(opts_.channels->begin(), opts_.channels->end(), key);
-      if (iter == opts_.channels->end()) {
-        continue;
+  auto registerChannel = [this](
+      std::vector<std::string>&& keys,
+      const std::function<void(int64_t, std::string&&)>& cb) {
+    for (auto& key : keys) {
+      int64_t priority = -1;
+      if (opts_.channels.has_value()) {
+        auto iter =
+            std::find(opts_.channels->begin(), opts_.channels->end(), key);
+        if (iter == opts_.channels->end()) {
+          continue;
+        }
+        // Assign priorities in reverse order of occurrence in the vector, so that
+        // a channel that comes before another receives a higher priority.
+        priority = opts_.channels->size() - 1 - (iter - opts_.channels->begin());
       }
-      // Assign priorities in reverse order of occurrence in the vector, so that
-      // a channel that comes before another receives a higher priority.
-      priority = opts_.channels->size() - 1 - (iter - opts_.channels->begin());
+      cb(priority, std::move(key));
     }
-    std::unique_ptr<CpuChannelRegistration> reg =
-        TensorPipeCpuChannelRegistry()->Create(key);
-    if (priority == -1) {
-      priority = reg->priority;
-    }
-    context_->registerChannel(
-        priority, std::move(key), std::move(reg->channel));
-  }
+  };
+
+  registerChannel(
+      TensorPipeCpuChannelRegistry()->Keys(),
+      [this](int64_t priority, std::string&& key) {
+        std::unique_ptr<CpuChannelRegistration> reg =
+            TensorPipeCpuChannelRegistry()->Create(key);
+        if (priority == -1) {
+          priority = reg->priority;
+        }
+        context_->registerChannel(
+            priority, std::move(key), std::move(reg->channel));
+      }
+  );
 
-#if TENSORPIPE_USE_CUDA_IPC_CHANNEL
+#ifdef USE_CUDA_NOT_ROCM
 
-  for (auto& key : TensorPipeCudaChannelRegistry()->Keys()) {
-    std::unique_ptr<CudaChannelRegistration> reg =
-        TensorPipeCudaChannelRegistry()->Create(key);
-    context_->registerChannel(
-        reg->priority, std::move(key), std::move(reg->channel));
-  }
+  registerChannel(
+      TensorPipeCudaChannelRegistry()->Keys(),
+      [this](int64_t priority, std::string&& key) {
+        std::unique_ptr<CudaChannelRegistration> reg =
+            TensorPipeCudaChannelRegistry()->Create(key);
+        if (priority == -1) {
+          priority = reg->priority;
+        }
+        context_->registerChannel(
+            priority, std::move(key), std::move(reg->channel));
+      }
+  );
```

Comment:
Nit: If you want to merge these two codepaths even further, there's a trick to do so. In principle we're dealing with different types so we would need templates, but templated lambdas are only available in C++-20 (the syntax is `[captures] <typename T> (params) { ... }`). However C++-14 has something that is almost the same, called generic lambdas, which are those that use `auto` for one of their params.

Hence you could change the `registerChannel` lambda to accept a `auto& registry` param, so that you can call it twice, once with `TensorPipeCpuChannelRegistry()` and once with `TensorPipeCudaChannelRegistry()`. (Inside that lambda, then, you'd have to replace `std::unique_ptr<FooChannelRegistration>` with `auto`, which is a bit less readable, but maybe we can add a comment).

Reply:
great, thanks, let me try this!